### DISCUSSION TITLE
RDP mouse issue

### DISCUSSION THREAD

**Discussion by MexiKing76** [sipper69/Control3#2] [General]
First off, great job on building this. I have been looking for something like this for a while. The way I want to use this is over an RDP connection. When I I do that, the video and keyboard work fine. However, the mouse is totally out of sync. I was wondering if you have tested that or have suggestions on how to resolve this issue.

**Comment by sipper69** [comment]
You mean the pointer does not catch up, right? 
The stream of data is not processed fast enough. 

Initially I processed the movement on parallel threads, but still noticed it sometimes.

Later I dropped that and just process movement when ready.

Maybe just test if this solved your problem by installing the update via the Microsoft App store. If it does I'll provide the details.

**Comment by MexiKing76** [comment]
Thanks for the fast response. I made a video showing you the behavior. The video is taken from the laptop that I am RDPing into. What happens is when I barely move my mouse, it shoots all over the place within Control3 and is not usable. I hope that makes sense.

https://streamable.com/wnts00

**Reply by sipper69** [reply]
Interesting. I'll test the same.

So from laptop A I do a remote desktop session via RDP to  laptop B. From laptop B I manage Laptop C via  Control3? Correct? Just curious, but why such a set up?

Is the pointer all over the place? Or is it moving very fast?

**Reply by MexiKing76** [reply]
Yes, you have my scenario understood correctly.

I should explain my use case. I am a systems admin and I have some systems at remote locations that I need occasional access to. These systems are non-routable in an isolated network and I need to work on them, from time to time. So I don't want to invest in a Raritan and would like to use some old laptops that are lying around to do this. I have remote hands to connect the laptop to these systems. 

The pointer is very fast and jumps around within the Control3 window. I know for Raritan once you have the pointer in the window it stays in the window and you cannot exit without hitting a hotkey.

**Reply by sipper69** [reply]
I could reproduce!
RDP is actually a virtual desktop, not just remote control, so the low-level Windows calls behave differently.  

**What Control3 does** 
1. put the hosts mouse pointer "aside" and keep it there (via P/Invoke)
2. get the raw-data from the hosts mouse to send to the remote PC (via DirectX)

**But... the host is now the RDP virtual desktop**
1. it does not handle setting the mouse "aside" (pointer is on the local machine, not on the Control3 host) 
2. the mouse its raw-data is  **emulated** by RDP
a. movement is now related to the RDP screen instead of the physical mouse movement
b. movement value is not the raw data, emulated data is "different"

I think the only solution is using absolute mouse positioning instead of relative mouse positioning. This is actually way easier.
It was my initial implementation, but Linux does not support absolute mouse positioning (out of the box).

With absolute mouse positioning the host and remote mouse pointers may both be visible and must be "synced". 
No low-level calls are needed, only some positioning calculations.

**Comment by sipper69** [comment]
Added absolute mouse positioning, works really well within an RDP session. 
Could even disable the local mouse when working remote (works depending on the RDP client, native Windows client works perfect).
Only thing is that absolute positioning is not supported by Linux (maybe you should use VNC instead of RDP).

Put the update on the Microsoft App Store, let me know if it works for you.
